Make Obj_dir functions a little safer

rgrinberg opened the pull request:
The family of unsafe functions would return a file path even if it's not
supposed to exist. It was the responsibility of the caller to make sure
that these functions were called correctly. Instead, these functions now
make sure they are called with correct module/visibility/cm_kind
combinations.

aalekseyev commented on src/dune_rules/install_rules.ml:
```diff
@@ -136,24 +136,30 @@ end = struct
+            List.filter_map Ml_kind.all ~f:(fun ml_kind ->
+                Obj_dir.Module.cmt_file obj_dir m ~ml_kind)
+          in
+          let if_ b x =
```
The implementation of `if_` looks backwards, but tests keep failing even if that is fixed.

rgrinberg replied:
Thanks. Fixed.

aalekseyev commented on src/dune_rules/obj_dir.ml:
```diff
@@ -293,27 +293,35 @@ module Module = struct
-    let obj_name = Module.obj_name m in
-    let fname = Module_name.Unique.artifact_filename obj_name ~ext in
-    relative t base fname
+  let cm_file_unsafe t m ~kind =
```
Since the function now makes the relevant check and raises if it's violated, we should call it `cm_file_exn`, not `cm_file_unsafe`.

aalekseyev commented:
I thought I'd find and fix the reason tests are failing, but it seems there're may be multiple reasons:
- `visibility` changed
- the ".o" file became unconditional in `other_cm_files` in install_rules.ml
- something else that I couldn't find

I suggest you make the changes in smaller steps.

rgrinberg commented:
>  I suggest you make the changes in smaller steps.

Thanks for debugging. I fixed these issues and reverted the visibility changes (will save those for another PR).

rgrinberg commented:
@aalekseyev small ping on this

aalekseyev commented:
The change looks good to me.

aalekseyev reviewed (approved):
Seems OK.
I do dislike all those new `_exn` occurrences, which make dune crash instead of possibly harmless behavior (lazy errors about no rule for ".cmx", for example). I did go through many of them, though, and they look benign.
